### PR TITLE
Graceful shutdown

### PR DESCRIPTION
**Pull request by changmyeong**

### CODE REVIEW

**Comment by changmyeong** [comment]
@kibertoad 

https://github.com/node-schedule/node-schedule/issues/549
PR on this issue.

**Comment by kibertoad** [comment]
@changmyeong Could you please add test for this?

**Comment by changmyeong** [comment]
@kibertoad I just added test code.

File: lib/Job.js [hunk 1]
```diff
@@ -41,6 +41,21 @@ function Job(name, job, callback) {
     this.callback = typeof callback === 'function' ? callback : false;
   }
 
+  // task count
+  this.running = 0;
+
+  this.on('run', function () {
```
Comment (kibertoad) [inline lib/Job.js:47]: I'm slightly concerned about generating three listeners per each job and then never removing them. Isn't that going to cause memory leaks, especially when jobs are being created dynamically?
Is there no place in the code where we could manage these counters without using listeners?

**Comment by kibertoad** [comment]
There's still a minor merge conflict remaining, but other than that, LGTM!

File: lib/schedule.js [hunk 2]
```diff
@@ -73,10 +73,42 @@ function cancelJob(job) {
   return success;
 }
 
+function gracefulShutdown() {
```
Comment (kibertoad) [inline lib/schedule.js:76]: Wouldn't it be more flexible if it didn't register itself on SIGINT, but instead provided a synchronous method that would gracefully stop schedule, which could be plugged by end user in whatever way they deem necessary, e. g. via https://github.com/mcollina/on-exit-leak-free or by using DI engine teardown mechanism?

File: lib/schedule.js [hunk 3]
```diff
@@ -73,10 +73,42 @@ function cancelJob(job) {
...
+
+    if (running) {
+      setInterval(function () {
+        for (let i = 0; i < jobs.length; i++) {
+          if (jobs[i].running > 0) {
+            return;
+          }
+        }
+  
+        process.exit(0);
```
Comment (kibertoad) [inline lib/schedule.js:99]: would be better if it was a graceful shutdown just of node-schedule, not entire Node process

Comment (kibertoad) [reply lib/schedule.js:99]: This also unlikely to work properly in browser

File: lib/Invocation.js [hunk 4]
```diff
@@ -267,19 +267,23 @@ function prepareNextInvocation() {
       try {
         const result = job.invoke(cinv.fireDate instanceof CronDate ? cinv.fireDate.toDate() : cinv.fireDate);
         job.emit('run');
+        job.running += 1;
 
         if (result instanceof Promise) {
-          result.then(function(value) {
+          result.then(function () {
```
Comment (kibertoad) [inline lib/Invocation.js:273]: value is lost here

Comment (changmyeong) [reply lib/Invocation.js]: It was an old PR, so there was a little difference. just solved it

File: lib/schedule.js (hunk 3 above)
Comment (changmyeong) [reply lib/schedule.js:99]: I never thought that this would work in browsers. Is there a good way to just shutdown node-schedule?

Comment (kibertoad) [reply lib/schedule.js:99]: Can't you just remove the SIGINT wrapping part of the `gracefulShutdown` function and make it do what SIGINT callback is currently doing? That way whoever will invoke `gracefulShutdown()` will effectively get node-schedule stopped after a while. Better yet, you can return a promise that would resolve at the point where you do process.exit(0) now.

Comment (changmyeong) [reply lib/schedule.js:99]: I thought of terminating the process after processing all the jobs when the system sends an interrupt signal. If you remove the SIGINT wrapping part, you will not be able to handle the interrupt signal event when it occurs.

Comment (kibertoad) [reply lib/schedule.js:99]: What I mean is people could do
```
process.on('SIGINT', async function () { 
  schedule.gracefulShutdown()
  .then(() => { process.exit(0) })
}
```
when running on node, and just
```
await schedule.gracefulShutdown()
```
on browser, if that's how they want to consume it.

And it probably should be documented.

Comment (changmyeong) [reply lib/schedule.js:99]: Great Idea! I'll do that.

**Comment by changmyeong** [comment]
@kibertoad I just deleted interrupt listener.

File: test/graceful-shutdown-test.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  t.test('Setup', function (test) {
+    clock = sinon.useFakeTimers();
+    sandbox = sinon.createSandbox({ useFakeTimers: true });
+    exitStub = sandbox.stub(process, 'exit');
+    schedule.gracefulShutdown();
+    test.end();
+  });
+
+  t.test('close immediately when no job', function (test) {
+    schedule.gracefulShutdown().then(function() {
```
Comment (kibertoad) [inline test/graceful-shutdown-test.js:18]: what is the effect if `gracefulShutdown` is an empty method? Would it hang?

File: test/graceful-shutdown-test.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+const sinon = require('sinon');
+const schedule = require('..');
+
+test('Graceful Shutdown', function (t) {
+  let clock, sandbox, exitStub;
+  t.test('Setup', function (test) {
+    clock = sinon.useFakeTimers();
+    sandbox = sinon.createSandbox({ useFakeTimers: true });
+    exitStub = sandbox.stub(process, 'exit');
+    schedule.gracefulShutdown();
```
Comment (kibertoad) [inline test/graceful-shutdown-test.js:13]: should it be part of the setup?

File: test/graceful-shutdown-test.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+const test = require('tape');
+const sinon = require('sinon');
+const schedule = require('..');
+
+test('Graceful Shutdown', function (t) {
+  let clock, sandbox, exitStub;
+  t.test('Setup', function (test) {
+    clock = sinon.useFakeTimers();
+    sandbox = sinon.createSandbox({ useFakeTimers: true });
+    exitStub = sandbox.stub(process, 'exit');
```
Comment (kibertoad) [inline test/graceful-shutdown-test.js:12]: is this still necessary?

File: lib/schedule.js (hunk 2 above)
Comment (kibertoad) [inline lib/schedule.js:76]: Can you add example to documentation of how to wrap it into SIGINT listener?

Comment (kibertoad) [reply lib/schedule.js:76]: oh, sorry, it's already there

**Comment by kibertoad** [comment]
Tests are failing on Node 6 because Object.values is not available there. Wonder if implementation could be adjusted to use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyNames instead.

File: test/graceful-shutdown-test.js (hunk 7 above)
Comment (kibertoad) [reply test/graceful-shutdown-test.js:12]: This one is still there

File: test/graceful-shutdown-test.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+    job.running = 0;
+    schedule.gracefulShutdown().then(function() {
+      test.ok(true);
+      test.end();
+    });
+  });
+
+  t.test('Restore', function (test) {
+    clock.restore();
```
Comment (kibertoad) [inline test/graceful-shutdown-test.js:32]: is it still needed

File: test/graceful-shutdown-test.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
+'use strict';
+
+const test = require('tape');
+const sinon = require('sinon');
```
Comment (kibertoad) [inline test/graceful-shutdown-test.js:4]: is sinon still needed?

File: test/graceful-shutdown-test.js (hunk 5 above)
Comment (changmyeong) [reply test/graceful-shutdown-test.js:9]: No. Promise will be returned, but there will be no delay.

Comment (kibertoad) [reply test/graceful-shutdown-test.js:8]: Is this test actually testing anything then? Should we at least assert amount of active jobs at its end then?

Comment (changmyeong) [reply test/graceful-shutdown-test.js:8]: It doesn't seem to make any sense after removing the SIGINT callback. I will delete that test case.

File: lib/schedule.js (hunk 3 above)
Comment (changmyeong) [reply lib/schedule.js]: Added a separate example to the document that can gracefully terminate with a SIGINT callback.

**Review by kibertoad — APPROVED** [review]

**Comment by kibertoad** [comment]
Thank you!

**Pull request merged**